auth.go

```diff
@@ -459,3 +460,41 @@ func (h *LogoutHandler) redirect(w http.ResponseWriter, r *http.Request, idToken
 
 	http.Redirect(w, r, redirectURL.String(), http.StatusFound)
 }
+
+// Handles token refreshes.
+type RefreshHandler struct {
+	auth     *Auth
+	keychain *keychain.Keychain
+}
+
+func newRefreshHandler(auth *Auth, keychain *keychain.Keychain) http.Handler {
+	return &RefreshHandler{auth, keychain}
+}
+
+func (h *RefreshHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !h.keychain.Guard(w, r) { // API
+		return
+	}
+
+	sessionID := r.Header.Get("Wave-Session-ID")
+	session, ok := h.auth.get(sessionID)
+
+	if !ok {
+		echo(Log{"t": "refresh_session", "error": "session unavailable"})
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+
+	token, err := h.auth.ensureValidOAuth2Token(r.Context(), session.token)
+	if err != nil {
+		// Purge session and reload clients if refresh not successful?
+		echo(Log{"t": "refresh_session", "error": err.Error()})
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+
+	session.token = token
```

Comment:
I might be wrong but I'd expect to call a `h.auth.set` after changing the session after modifying it. Is it missing here?

Reply:
Good point, done.